perf.go

```diff
@@ -283,33 +305,53 @@ func (pr *PerfReader) LostSamples() uint64 {
 	return atomic.LoadUint64(&pr.lostSamples)
 }
 
-// Close stops the reader.
+// Close stops the reader, discarding any samples not yet written to 'Samples'.
 //
 // Calls to perf_event_output from eBPF programs will return
 // ENOENT after calling this method.
 func (pr *PerfReader) Close() (err error) {
+	// Interrupt polling so we don't deadlock if the consumer is dead
+	close(pr.stopWriter)
+
+	return pr.close(pr.closeFd)
+}
+
+// Close stops the reader, flushing any samples to 'Samples'.
+// Will block if no consumer reads from 'Samples'.
+//
+// Calls to perf_event_output from eBPF programs will return
+// ENOENT after calling this method.
+func (pr *PerfReader) FlushAndClose() error {
+	return pr.close(pr.flushCloseFd)
+}
+
+func (pr *PerfReader) close(eventFd int) error {
 	pr.closeOnce.Do(func() {
 		runtime.SetFinalizer(pr, nil)
 
-		// Indicate that we want to shut down
-		close(pr.close)
 		pr.array.Close()
 
 		// Signal poll() via the event fd. Ignore the
 		// write error since poll() may have exited
 		// and closed the fd already
 		var value [8]byte
 		nativeEndian.PutUint64(value[:], 1)
-		_, _ = syscall.Write(pr.closeFd, value[:])
+		_, _ = syscall.Write(eventFd, value[:])
 	})
 
+	pr.closed.Wait()
```

Comment:
This Wait() and Add(1) from poll() can race, which is UB https://golang.org/pkg/sync/#WaitGroup.Add

I think it's better to read from another `chan struct` here.